src/Database/PG/Query/Connection.hs

```diff
@@ -404,12 +404,21 @@ cancelPG c = do
 -- The caller can't tell whether a transaction was committed.
 cancelOnAsync :: PQ.Connection -> IO a -> PGExec a
 cancelOnAsync conn action = do
-  c <-
-    lift (PQ.getCancel conn) >>= \case
-      Nothing -> throwPGIntErr $ PGIUnexpected $ "failed to allocate cancel handle"
-      Just c -> pure c
-  lift (interruptOnAsyncException (cancelPG c) action)
-    `catch` (\(PGCancelErr msg) -> throwPGIntErr $ PGIUnexpected $ "error cancelling query: " <> msg)
+  lift (PQ.getCancel conn) >>= \case
+    -- We can't get a cancel handle when the connection is invalid [1]. So we
+    -- don't setup the cancellation, because the query is going to fail anyway
+    -- on this invalid connection. If we throw an error here, it becomes
+    -- confusing for the user. An appropriate way to handle it would be to throw
+    -- an invalid connection error here. But we don't have any details on why
+    -- the connection is invalid. If we let postgres throw the error, it will
+    -- have the actual error message/reason for the invalid connection. Hence,
+    -- it makes sense to not setup the cancellation here and try to run the
+    -- query, which will fail with proper error message from Postgres.
+    -- [1]: https://hackage.haskell.org/package/postgresql-libpq-0.9.4.3/docs/Database-PostgreSQL-LibPQ.html#v:getCancel
```

Comment:
I was initially confused about some minor things here:

1. If a connection is invalid, then why did connecting fail? I think it'd help to phrase this as: "when the connection _has become_ invalid". Every `PGConn` represents what was _once_ a functioning connection to Postgres, but this may have changed.
2. How can a connection suddenly become invalid? Didn't we just connect to Postgres? In fact, the cancel handle is setup when the connection is acquired through the pool, which may be much later than when the connection is actually setup. So it may be e.g. several minutes or hours old already. So the confusion here is that a cancel handle isn't created when the connection is established, but when the connection is lent out through the pool.
3. If we don't throw an error here, what will? I.e. what does "let postgres throw the error" refer to? Well, when we'd try to execute actual SQL statements, `libpq` would find out that the connection has become invalid.
4. (Not sure this is worth mentioning.) Is it possible that a connection is invalid when we attempt to create the cancel handle, but later becomes valid again? I'm not sure. But in any case, the only effect of that would be that we wouldn't catch any async exceptions, i.e. a timeout wouldn't trigger a PG operation cancellation.

Reply:
> Is it possible that a connection is invalid when we attempt to create the cancel handle, but later becomes valid again?
That's a good question, and I couldn't find/think of any edge-case where this can happen.

Reply:
@abooij I've updated the comment based on your comments here. Do you want to re-review it?